### PR TITLE
Feature / HOLO 1531 Add HydrateNFT to SDK

### PR DESCRIPTION
**Pull request by BravoNatalie**
## Describe Changes

Changes made to the package: **sdk**.

I made it better by doing ...

- Add `hydrateNFT` to `HolographProtocol` service
- Add `hydrateNFT` tests
- Change `OpenEditionNFT` class to better represent the on-chain data

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] Code styles have been enforced
- [x] All tests are passing

### CODE REVIEW

File: packages/sdk/src/assets/nft.validation.ts [hunk 1]
```diff
@@ -63,20 +63,13 @@ export const validate = {
 }
 
 export type CreateNFTSchema = z.infer<typeof createNFTSchema>
+export type NFTMetadata = z.infer<typeof metadataSchema>
+export type IpfsInfo = z.infer<typeof ipfsInfoSchema>
 
 export type CreateNFT = {
   contract: HolographERC721Contract | HolographOpenEditionERC721ContractV1 | HolographOpenEditionERC721ContractV2
-  ipfsInfo?: {
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/nft.validation.ts:69]: You'll need to revert these changes, this will mess up with the TypeScript intellisense and the user will not be able to get TS autocomplete from this properties unless the user also installs Zod

File: packages/sdk/src/services/holograph-protocol.service.ts [hunk 2]
```diff
@@ -204,4 +207,80 @@ export class HolographProtocol {
         throw new Error('This type of contract is not currently supported.')
     }
   }
+
+  async hydrateNFT(hydrateNftInput: {
```
Comment (feliperodriguess) [inline packages/sdk/src/services/holograph-protocol.service.ts:211]: Change this to `hydrateNFTFromTokenId` to keep the pattern

File: packages/sdk/src/services/holograph-protocol.service.ts [hunk 3]
```diff
@@ -204,4 +207,80 @@ export class HolographProtocol {
...
+    chainId: number
+    address: Address
```
Comment (feliperodriguess) [inline packages/sdk/src/services/holograph-protocol.service.ts:213]: I think calling this as `contractAddress` will make things more clear to the users

File: packages/sdk/src/errors/assets/token-does-not-exist.error.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import {HolographError, HolographErrorCode} from '../holograph-error'
+
+export const TOKEN_ID_DOES_NOT_EXIST_ERROR_MESSAGE = 'ERC721: token ID does not exist'
```
Comment (feliperodriguess) [inline packages/sdk/src/errors/assets/token-does-not-exist.error.ts:3]: Maybe replace ERC721 with HOLOGRAPH (keep same pattern)

File: packages/sdk/src/tests/assets/open-edition-nft.test.ts [hunk 5]
```diff
@@ -48,35 +48,18 @@ describe('Asset class: OpenEditionNFT', () => {
...
-        attributes: {
-          key: 'value',
-          anotherKey: 'anotherValue',
-        },
-      },
     })
   })
 
-  it('should be able to get the OpenEditionNFT wrapper class', () => {
+  it.skip('should be able to get the OpenEditionNFT wrapper class', () => {
```
Comment (feliperodriguess) [inline packages/sdk/src/tests/assets/open-edition-nft.test.ts:54]: i think we need to add this back

File: packages/sdk/src/tests/assets/open-edition-nft.test.ts [hunk 6]
```diff
@@ -48,35 +48,18 @@ describe('Asset class: OpenEditionNFT', () => {
...
-    expect(nft).toHaveProperty('attributes')
-    expect(nft).toHaveProperty('ipfsUrl')
+    expect(nft).toHaveProperty('properties')
     expect(nft).toHaveProperty('ipfsImageCid')
+    expect(nft).toHaveProperty('animationUrl')
     expect(nft).toHaveProperty('getOwner')
     expect(nft).toHaveProperty('getParsedTokenId')
-    expect(nft).toHaveProperty('setMetadata')
-    expect(nft).toHaveProperty('setName')
-    expect(nft).toHaveProperty('setDescription')
```
Comment (feliperodriguess) [inline packages/sdk/src/tests/assets/open-edition-nft.test.ts:75]: why remove all of those methods here if they are still present in the class?

File: packages/sdk/src/assets/open-edition-nft.ts [hunk 7]
```diff
@@ -1,20 +1,73 @@
...
+    this.holographOpenEditionERC721 = new HolographOpenEditionERC721Contract(contract.contractAddress!, version)
+    this.isMinted = false
+  }
+
+  @RequireMintedToken()
+  get metadata() {
+    return this._metadata
+  }
+
+  @RequireMintedToken()
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/open-edition-nft.ts:38]: Could you let me know why we need this here? 

I think if I just instantiated the class, I should be able to get my NFT name even though I haven't minted yet

File: packages/sdk/src/assets/open-edition-nft.ts [hunk 8]
```diff
@@ -1,20 +1,73 @@
...
+  get name() {
+    return this._metadata?.name
+  }
+
+  @RequireMintedToken()
+  get description() {
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/open-edition-nft.ts:44]: Same for all of these info and beyond

File: packages/sdk/src/assets/open-edition-nft.ts [hunk 9]
```diff
@@ -1,20 +1,73 @@
...
+import {CreateOpenEditionNFT, HolographOpenEditionNFTMetadata, validate} from './nft.validation'
+import {RequireMintedToken} from '../utils/decorators'
+import {getParsedTokenId} from '../utils/transformers'
+import {
+  HolographOpenEditionERC721ContractV1,
+  HolographOpenEditionERC721ContractV2,
+} from './holograph-open-edition-erc721-contract'
+import {NotMintedNFTError} from '../errors'
+
+export class OpenEditionNFT {
```
Comment (feliperodriguess) [inline packages/sdk/src/assets/open-edition-nft.ts:16]: We are not extending the NFT class anymore, why this change?

**Review by feliperodriguess — CHANGES_REQUESTED** [review]

File: packages/sdk/src/tests/assets/open-edition-nft.test.ts [hunk 10]
```diff
@@ -48,35 +48,18 @@ describe('Asset class: OpenEditionNFT', () => {
...
     expect(nft).toHaveProperty('description')
-    expect(nft).toHaveProperty('creator')
-    expect(nft).toHaveProperty('attributes')
-    expect(nft).toHaveProperty('ipfsUrl')
+    expect(nft).toHaveProperty('properties')
     expect(nft).toHaveProperty('ipfsImageCid')
+    expect(nft).toHaveProperty('animationUrl')
     expect(nft).toHaveProperty('getOwner')
     expect(nft).toHaveProperty('getParsedTokenId')
-    expect(nft).toHaveProperty('setMetadata')
```
Comment (feliperodriguess) [inline packages/sdk/src/tests/assets/open-edition-nft.test.ts:73]: we need these kind of methods back i guess

File: packages/sdk/src/assets/open-edition-nft.ts (hunk 7 above)
Comment (BravoNatalie) [reply packages/sdk/src/assets/open-edition-nft.ts:38]: The NFT Open Edition (OE) name follows this pattern: CollectionName + TokenID. Since the TokenID is only generated after the token is minted, we can't determine the name beforehand.

File: packages/sdk/src/assets/open-edition-nft.ts (hunk 8 above)
Comment (BravoNatalie) [reply packages/sdk/src/assets/open-edition-nft.ts:44]: The description can be obtained before the token is minted, as it matches the collection's description. However, I followed the pattern where all this data comes from the TokenURI, which is only generated after minting.

File: packages/sdk/src/tests/assets/open-edition-nft.test.ts (hunk 10 above)
Comment (BravoNatalie) [reply packages/sdk/src/tests/assets/open-edition-nft.test.ts:73]: We can't set the metadata for an OE NFT, it's defined when we deploy the OE collection. The NFT metadata is then derived from the collection's metadata.

File: packages/sdk/src/assets/open-edition-nft.ts (hunk 9 above)
Comment (BravoNatalie) [reply packages/sdk/src/assets/open-edition-nft.ts:16]: This is due to the way metadata works. For standard NFTs, we can set individual metadata, but this isn't the case for Open Edition (OE) NFTs. Besides that, some metadata properties differ between standard NFTs and OE NFTs.

**Review by feliperodriguess — APPROVED** [review]

**Review by sogoiii — APPROVED** [review]
👍
